### PR TITLE
Create crossword_puzzle_solver.py

### PR DESCRIPTION
**Pull request by Khushi-Shukla**
### Describe your change:
Added an algorithm in the backtracking folder

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Add or change doctests? -- Note: Please avoid changing both code and tests in a single pull request.
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [ ] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

### CODE REVIEW

File: backtracking/crossword_puzzle_solver.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
+# https://www.geeksforgeeks.org/solve-crossword-puzzle/
+from typing import List, Optional
```
Comment (cclauss) [inline backtracking/crossword_puzzle_solver.py:2]: These are not needed in Python 3.12

File: backtracking/crossword_puzzle_solver.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+
+def solve_crossword(
+    puzzle: List[List[str]], words: List[str]
```
Comment (cclauss) [inline backtracking/crossword_puzzle_solver.py:6]:
```suggestion
    puzzle: list[list[str]], words: list[str]
```

**Comment by Khushi-Shukla** [comment]
Changes have been made. Please review the PR @cclauss

**Comment by cclauss** [comment]
I will look tonight.

**Comment by Khushi-Shukla** [comment]
Okay

File: backtracking/crossword_puzzle_solver.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    ...     ['', '', '', ''],
+    ...     ['', '', '', ''],
+    ...     ['', '', '', ''],
+    ...     ['', '', '', '']
+    ... ]
+
+    >>> words = ['word', 'four', 'more', 'last']
+    >>> solve_crossword(puzzle, words)
+    True
+    """
```
Comment (cclauss) [inline backtracking/crossword_puzzle_solver.py:73]: Should we be able to solve this crossword puzzle?
```suggestion
    >>> words = ['word', 'four', 'more', 'paragraphs']
    >>> solve_crossword(puzzle, words)
    False
    """
```

File: backtracking/crossword_puzzle_solver.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    ... ]
+    >>> place_word(puzzle, 'word', 0, 0, True)
+    >>> puzzle
+    [['w', '', '', ''], ['o', '', '', ''], ['r', '', '', ''], ['d', '', '', '']]
+    """
+    for i in range(len(word)):
+        if vertical:
+            puzzle[row + i][col] = word[i]
+        else:
+            puzzle[row][col + i] = word[i]
```
Comment (cclauss) [inline backtracking/crossword_puzzle_solver.py:57]:
```suggestion
    for i, char in enumerate(word):
        if vertical:
            puzzle[row + i][col] = char
        else:
            puzzle[row][col + i] = char
```

File: backtracking/crossword_puzzle_solver.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+                            if solve_crossword(puzzle, words):
+                                return True
+                            words.append(word)
+                            remove_word(puzzle, word, row, col, vertical)
+                return False
+    return True
+
+
+PUZZLE = [["" for _ in range(3)] for _ in range(3)]
+WORDS = ["cat", "dog", "car"]
```
Comment (cclauss) [inline backtracking/crossword_puzzle_solver.py:124]:
```suggestion
PUZZLE = [[""] * 3] for _ in range(3)]
WORDS = ("cat", "dog", "car")
```

File: backtracking/crossword_puzzle_solver.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+if solve_crossword(PUZZLE, WORDS):
+    print("Solution found:")
+    for row in PUZZLE:
+        print(" ".join(row))
+else:
+    print("No solution found:")
```
Comment (cclauss) [inline backtracking/crossword_puzzle_solver.py:131]:
```suggestion
if __name__ == "__main__":
    PUZZLE = [[""] * 3] for _ in range(3)]
    WORDS = ("cat", "dog", "car")

    if solve_crossword(PUZZLE, WORDS):
        print("Solution found:")
        for row in PUZZLE:
            print(" ".join(row))
    else:
        print("No solution found:")
```

**Review by cclauss — APPROVED** [review]

**Pull request merged**